rocAL/rocAL/source/video_read_and_decode.cpp

```diff
@@ -0,0 +1,292 @@
+/*
+Copyright (c) 2019 - 2020 Advanced Micro Devices, Inc. All rights reserved.
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+*/
+
+#include "video_decoder_factory.h"
+#include "video_read_and_decode.h"
+
+namespace filesys = boost::filesystem;
+
+#ifdef RALI_VIDEO
+std::tuple<VideoDecoder::ColorFormat, unsigned, AVPixelFormat>
+video_interpret_color_format(RaliColorFormat color_format)
+{
+    switch (color_format)
+    {
+        case RaliColorFormat::RGB24:
+            return std::make_tuple(VideoDecoder::ColorFormat::RGB, 3, AV_PIX_FMT_RGB24);
+
+        case RaliColorFormat::BGR24:
+            return std::make_tuple(VideoDecoder::ColorFormat::BGR, 3, AV_PIX_FMT_BGR24);
+
+        case RaliColorFormat::U8:
+            return std::make_tuple(VideoDecoder::ColorFormat::GRAY, 1, AV_PIX_FMT_GRAY8);
+
+        default:
+            throw std::invalid_argument("Invalid color format\n");
+    }
+}
+
+Timing
+VideoReadAndDecode::timing()
+{
+    Timing t;
+    t.image_decode_time = _decode_time.get_timing();
+    t.image_read_time = _file_load_time.get_timing();
+    t.shuffle_time = _reader->get_shuffle_time();
+    return t;
+}
+
+VideoReadAndDecode::VideoReadAndDecode() : _file_load_time("FileLoadTime", DBG_TIMING),
+                                           _decode_time("DecodeTime", DBG_TIMING)
+{
+}
+
+VideoReadAndDecode::~VideoReadAndDecode()
+{
+    _reader = nullptr;
+    _video_decoder.clear();
+}
+
+void VideoReadAndDecode::create(ReaderConfig reader_config, VideoDecoderConfig decoder_config, int batch_size)
+{
+
+    _sequence_length = reader_config.get_sequence_length();
+    _stride = reader_config.get_frame_stride();
+    _video_count = reader_config.get_video_count();
+    _video_names = reader_config.get_video_file_names();
+    _frame_rate = reader_config.get_video_frame_rate();
+    _batch_size = batch_size;
+    set_video_process_count(_video_count);
```

Comment:
we need to only decode a batch of video sequences at a time. So the video thread count should be equal to the number of separate videos in a batch. Setting it equal to available videos is not a good idea since it can be huge

Reply:
We have discussed this in our earlier meetings and we decided to set a upper limit of the _video_process_count (_max_video_count = 50).  This will ensure at the max we will create video instances for 50 or less videos.